src/AzureExtension/DevBox/DevBoxProvider.cs

```diff
@@ -143,6 +143,15 @@ public async Task<IEnumerable<IComputeSystem>> GetDevBoxesAsync(IDeveloperId dev
         // update the cache every time we retrieve new Dev Boxes. This is used so in the creation flow we don't need
         // to retrieve the Dev Boxes again if the user already retrieved them in the environments page in Dev Home
         _cachedDevBoxesMap[GetUniqueDeveloperId(developerId)] = devBoxes.ToList();
+
+        // Get Associated pools the first time we get the projects
+        var uniqueUserId = GetUniqueDeveloperId(developerId);
+        if (!_devBoxProjectAndPoolsMap.TryGetValue(uniqueUserId, out var _))
+        {
+            _log.Information($"Found no cached pools for all projects for {developerId.LoginId}, retrieving pools");
+            _ = Task.Run(async () => _devBoxProjectAndPoolsMap[uniqueUserId] = await _devBoxManagementService.GetAllProjectsToPoolsMappingAsync(devBoxProjects!, developerId));

```

Comment:
Only Issue I see here now that this is being done in another thread:

1. User launches Dev Home
2. User goes to Machine Configuration
3. User goes to Create Virtual Environment
4. User selects Dev Box
5. User clicks next.

Do the pools show up? I don't think they do as they'll still be loading, I think we'll need a lock or an async semaphore that only allows 1 entry at a time in line 152 
 that covers   this part `_devBoxProjectAndPoolsMap[uniqueUserId] = await _devBoxManagementService.GetAllProjectsToPoolsMappingAsync(devBoxProjects!, developerId))`


then in the `CreateAdaptiveCardSessionForDeveloperId` method here: 
https://github.com/microsoft/DevHomeAzureExtension/blob/13bed06622b878023931a666471581e61e7bb96c/src/AzureExtension/DevBox/DevBoxProvider.cs#L227

use it to block execution until the semphore (can use an Auto or Manual event too) above is exited.

Reply:
I completely missed the flow from the machine config; I just tested with the Environments tab in mind.
Also, I moved this to a background thread to speed things up as this was one of the biggest reasons for the slow loading. So I have moved this out now. Let me know if you have any reservations.